### PR TITLE
Schema compliance

### PR DESCRIPTION
**Pull request by luiztauffer**
Necessary changes for compliance with input and metadata schemas and manipulation.

### CODE REVIEW

File: nwb_conversion_tools/basedatainterface.py [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
+        return input_schema
 
     def __init__(self, **input_args):
         self.input_args = input_args
 
-    @abstractmethod
     def get_metadata_schema():
-        pass
+        metadata_schema = get_metadata_schema()
+        return metadata_schema
```
Comment (bendichter) [inline nwb_conversion_tools/basedatainterface.py]: @luiztauffer What's going on here? The function is calling itself?

Comment (luiztauffer) [reply nwb_conversion_tools/basedatainterface.py]: no, its calling get_metadata_schema from utils, which gives a basic, blank metadata_schema

File: nwb_conversion_tools/basedatainterface.py [hunk 2]
```diff
@@ -1,20 +1,21 @@
...
 
     def __init__(self, **input_args):
         self.input_args = input_args
 
-    @abstractmethod
-    def get_metadata_schema(self):
-        pass
+    def get_metadata_schema():
+        metadata_schema = get_metadata_schema()
+        return metadata_schema
```
Comment (bendichter) [inline nwb_conversion_tools/basedatainterface.py:18]:
```suggestion
        return get_metadata_schema()
```

File: nwb_conversion_tools/nwbconverter.py [hunk 3]
```diff
@@ -30,19 +33,22 @@ def __init__(self, **input_data):
...
-                k: input_data[interface_name].get(k, None)
-                for k in interface_schema['properties'].keys()
-            }
+            input_data_routed[interface_name] = dict()
+            interface_schema_properties = interface.get_input_schema()['properties']
+            for b in set(interface_schema_properties.keys()).intersection(set(['source_data', 'conversion_options'])):
+                input_data_routed[interface_name].update({
+                    k: input_data[b].get(k, None)
+                    for k in interface_schema_properties[b]['properties'].keys()
+                })
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:42]:
```suggestion
            interface_schema_properties = interface.get_input_schema()['properties']
            for b in set(interface_schema_properties.keys()).intersection(set(['source_data', 'conversion_options'])):
                input_data_routed[interface_name] = {
                    k: input_data[b].get(k, None)
                    for k in interface_schema_properties[b]['properties'].keys()
                }
```

File: nwb_conversion_tools/nwbconverter.py [hunk 4]
```diff
@@ -62,24 +68,33 @@ def get_metadata(self):
...
         """Build nwbfile object, auto-populate with minimal values if missing."""
+
+        # Minimal values
         nwbfile_kwargs = dict(
-                session_description="no description",
-                identifier=str(uuid.uuid4()),
-                session_start_time=datetime.now()
+            session_description="no description",
+            identifier=str(uuid.uuid4()),
+            session_start_time=datetime.now()
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:76]:
```suggestion
```

File: nwb_conversion_tools/nwbconverter.py [hunk 5]
```diff
@@ -62,24 +68,33 @@ def get_metadata(self):
...
-            nwbfile_kwargs.update(subject=Subject(**metadata_dict['Subject']))
-
+        # Initiate nwbfile
+        nwbfile_kwargs.update(metadata_dict['NWBFile'])
         nwbfile = NWBFile(**nwbfile_kwargs)
 
+        # add Subject
+        if 'Subject' not in metadata_dict:
+            metadata_dict['Subject'] = dict()
+        nwbfile.subject = Subject(**metadata_dict['Subject'])
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:91]: We shouldn't automatically create an empty Subject

File: nwb_conversion_tools/baserecordingextractorinterface.py [hunk 6]
```diff
@@ -20,16 +20,17 @@ def __init__(self, **input_args):
...
+        # Initiate Ecephys metadata
+        metadata_schema['properties']['Ecephys'] = dict()
+        metadata_schema['properties']['Ecephys'].update(
             Device=get_schema_from_hdmf_class(Device),
             ElectrodeGroup=get_schema_from_hdmf_class(ElectrodeGroup),
             ElectricalSeries=get_schema_from_hdmf_class(ElectricalSeries)
         )
-        required_fields = ['Device', 'ElectrodeGroup', 'ElectricalSeries']
-        metadata_schema['required'] += required_fields
+        metadata_schema['properties']['Ecephys'] = ['Device', 'ElectrodeGroup', 'ElectricalSeries']
```
Comment (luiztauffer) [inline nwb_conversion_tools/baserecordingextractorinterface.py:33]: @CodyCBakerPhD could you please check this? Its likely important for the Converters using BaseRecordingExtractorInterface

**Comment by luiztauffer** [comment]
https://github.com/catalystneuro/nwb-conversion-tools/blob/23fafea7b9b5430c4fba2764f4b073cd797c0eb5/nwb_conversion_tools/baserecordingextractorinterface.py#L50-L52

What is `subset_channels`? It feels like this shouldn't be on metadata
@CodyCBakerPhD

File: nwb_conversion_tools/nwbconverter.py (hunk 4 above)
Comment (luiztauffer) [reply nwb_conversion_tools/nwbconverter.py]: why remove only the default value for `session_start_time`? @bendichter

File: nwb_conversion_tools/nwbconverter.py (hunk 5 above)
Comment (luiztauffer) [reply nwb_conversion_tools/nwbconverter.py]: 820547d

File: nwb_conversion_tools/baserecordingextractorinterface.py [hunk 7]
```diff
@@ -20,16 +20,16 @@ def __init__(self, **input_args):
...
-        metadata_schema['properties'].update(
+        # Initiate Ecephys metadata
+        metadata_schema['properties']['Ecephys'] = dict(
             Device=get_schema_from_hdmf_class(Device),
             ElectrodeGroup=get_schema_from_hdmf_class(ElectrodeGroup),
             ElectricalSeries=get_schema_from_hdmf_class(ElectricalSeries)
         )
-        required_fields = ['Device', 'ElectrodeGroup', 'ElectricalSeries']
-        metadata_schema['required'] += required_fields
+        metadata_schema['properties']['Ecephys'] = ['Device', 'ElectrodeGroup', 'ElectricalSeries']
```
Comment (bendichter) [inline nwb_conversion_tools/baserecordingextractorinterface.py:32]:
```suggestion
```

Comment (bendichter) [inline nwb_conversion_tools/baserecordingextractorinterface.py:32]:
```suggestion
        metadata_schema['properties']['Ecephys']['required'] = ['Device', 'ElectrodeGroup', 'ElectricalSeries']
```

File: nwb_conversion_tools/nwbconverter.py [hunk 8]
```diff
@@ -15,23 +15,51 @@ class NWBConverter:
     data_interface_classes = None
 
     @classmethod
-    def get_input_schema(cls):
+    def get_source_schema(cls):
         """Compile input schemas from each of the data interface classes."""
         input_schema = get_input_schema()
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:20]:
```suggestion
        source_schema = get_source_schema()
```

File: nwb_conversion_tools/nwbconverter.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 """Authors: Cody Baker and Ben Dichter."""
-import uuid
-from datetime import datetime
-
+from .utils import (get_schema_from_hdmf_class, get_root_schema, get_input_schema,
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:2]:
```suggestion
from .utils import (get_schema_from_hdmf_class, get_root_schema, get_source_schema,
```

File: nwb_conversion_tools/nwbconverter.py [hunk 10]
```diff
@@ -15,23 +15,51 @@ class NWBConverter:
     data_interface_classes = None
 
     @classmethod
-    def get_input_schema(cls):
+    def get_source_schema(cls):
         """Compile input schemas from each of the data interface classes."""
         input_schema = get_input_schema()
         for name, data_interface in cls.data_interface_classes.items():
-            input_schema['properties'].update(name=data_interface.get_input_schema())
+            input_schema['properties'][name] = data_interface.get_input_schema()
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:22]:
```suggestion
            source_schema['properties'][name] = data_interface.get_source_schema()
```

File: nwb_conversion_tools/nwbconverter.py [hunk 11]
```diff
@@ -15,23 +15,51 @@ class NWBConverter:
...
+
         return input_schema
 
+    @classmethod
+    def get_conversion_options_schema(cls):
+        conversion_options_schema = get_base_schema()
+        for data_interface in cls.data_interface_classes.values():
+            conversion_options_schema['properties'].update(data_interface.get_conversion_options_schema()['properties'])
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:30]:
```suggestion
        for name, data_interface in cls.data_interface_classes.items()::
            conversion_options_schema['properties'][name] = data_interface.get_conversion_options_schema()
```

File: nwb_conversion_tools/nwbconverter.py [hunk 12]
```diff
@@ -15,23 +15,51 @@ class NWBConverter:
...
+
+        return conversion_options_schema
+
+    @classmethod
+    def get_input_schema(cls):
+        schema = get_root_schema()
+        schema['properties'].update(
+            source=cls.get_source_schema(),
+            conversion_options=cls.get_conversion_options_schema()
+        )
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:40]:
```suggestion
```

File: nwb_conversion_tools/nwbconverter.py [hunk 13]
```diff
@@ -15,23 +15,51 @@ class NWBConverter:
...
+            for b in set(interface_schema_properties.keys()).intersection(set(['source_data', 'conversion_options'])):
+                input_data_routed[interface_name] = {
+                    k: input_data[b].get(k, None)
+                    for k in interface_schema_properties[b]['properties'].keys()
+                }
+
+        self.data_interface_objects = {
+            name: data_interface(**input_data_routed[name])
+            for name, data_interface in self.data_interface_classes.items()
+        }
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:62]:
```suggestion
        self.data_interface_objects = {name: data_interface(**input_data[name])
                                       for name, data_interface in
                                       self.data_interface_classes.items()}
```

File: nwb_conversion_tools/nwbconverter.py [hunk 14]
```diff
@@ -61,26 +78,26 @@ def get_metadata(self):
...
-        if 'NWBFile' in metadata_dict:
-            nwbfile_kwargs.update(metadata_dict['NWBFile'])
-        if 'Subject' in metadata_dict:
-            nwbfile_kwargs.update(subject=Subject(**metadata_dict['Subject']))
+
+        # Minimal values
+        nwbfile_kwargs = dict(
+            session_description="no description",
+            identifier=str(uuid.uuid4()),
+        )
```
Comment (bendichter) [inline nwb_conversion_tools/nwbconverter.py:86]:
```suggestion
```

**Pull request merged**